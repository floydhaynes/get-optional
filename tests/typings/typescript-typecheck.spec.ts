import {execFile} from 'child_process';
import * as path from 'path';

const executableExtension = process.platform === 'win32' ? '.cmd' : '';

const checkTs = (testdir: string) => new Promise((resolve, reject) => execFile(
	path.join(ROOT_DIR, 'node_modules/.bin/tsc' + executableExtension),
	`--noEmit --project ./ts-${testdir}/tsconfig.json`.split(' '),
	{
		cwd: path.join(__dirname, 'testbed'),
	},
	(error, stdout, stderr) => {
		if (error) {
			reject(stdout || stderr || error);
		} else {
			resolve(stdout);
		}
	},
));

describe('Correct usage from TS', () => {
	it('should typecheck', () => {
		expect.assertions(1);
		const result = checkTs('should-pass');
		return expect(result).resolves.toBe('');
	});
});

describe('Incorrect usage from TS', () => {
	it('should not typecheck', () => {
		expect.assertions(1);
		const result = checkTs('should-not-pass');
		return expect(result).rejects.toMatchSnapshot();
	});
});
