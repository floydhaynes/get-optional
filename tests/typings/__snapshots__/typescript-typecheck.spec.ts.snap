// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typescript type definitions should not typecheck for incorrect usage 1`] = `
"ts-should-not-pass/get.ts(5,7): error TS2322: Type 'C | undefined' is not assignable to type 'C'.
  Type 'undefined' is not assignable to type 'C'.
ts-should-not-pass/get.ts(8,27): error TS2345: Argument of type '\\"c\\"' is not assignable to parameter of type '\\"b\\"'.
ts-should-not-pass/get.ts(11,28): error TS2345: Argument of type '\\"a\\"' is not assignable to parameter of type 'never'.
ts-should-not-pass/getWithDefault.ts(5,7): error TS2322: Type 'C | B' is not assignable to type 'C'.
  Type 'B' is not assignable to type 'C'.
    Property 'd' is missing in type 'B'.
ts-should-not-pass/getWithDefault.ts(9,48): error TS2345: Argument of type '\\"c\\"' is not assignable to parameter of type '\\"b\\"'.
ts-should-not-pass/nth.ts(9,7): error TS2322: Type 'boolean | undefined' is not assignable to type 'boolean'.
  Type 'undefined' is not assignable to type 'boolean'.
ts-should-not-pass/nth.ts(12,7): error TS2322: Type 'boolean | undefined' is not assignable to type 'string | undefined'.
  Type 'false' is not assignable to type 'string | undefined'.
ts-should-not-pass/nth.ts(15,24): error TS2345: Argument of type 'B | null | undefined' is not assignable to parameter of type 'ReadonlyArray<{}> | null | undefined'.
  Type 'B' is not assignable to type 'ReadonlyArray<{}>'.
    Property 'length' is missing in type 'B'.
ts-should-not-pass/nth.ts(18,40): error TS2345: Argument of type '\\"0\\"' is not assignable to parameter of type 'number'.
"
`;
